sklearn/conftest.py

```diff
@@ -227,3 +229,77 @@ def pytest_configure(config):
         matplotlib.use("agg")
     except ImportError:
         pass
+
+
+# Definition of the random_seed fixture. See the docstring of the fixture for
+# more information.
+
+RANDOM_SEED_RANGE = list(range(100))  # All seeds in [0, 99] should be valid.
+random_seed_var = environ.get("SKLEARN_TESTS_RANDOM_SEED")
+if random_seed_var is None:
+    # If the environment variable is not defined, pick-up one seed at random in
+    # the range of admissible random seeds. Note, to make sure that all
+    # pytest-xdist workers see the same seed, we seed the meta-random number
+    # generator with a value derived from the year and the day.
+    rng = Random(int(datetime.now().strftime("%Y%j")))
+    random_seeds = [rng.choice(RANDOM_SEED_RANGE)]
+elif random_seed_var == "all":
+    random_seeds = RANDOM_SEED_RANGE
+else:
+    if "-" in random_seed_var:
+        start, stop = random_seed_var.split("-")
+        random_seeds = list(range(int(start.strip()), int(stop.strip()) + 1))
+    else:
+        random_seeds = [int(s.strip()) for s in random_seed_var.split(",")]
+
+if min(random_seeds) < 0 or max(random_seeds) > 99:
+    raise ValueError(
+        "The value(s) of the environment variable SKLEARN_TESTS_RANDOM_SEED "
+        f"must be in the range [0, 99] (or 'all'), got: {random_seed_var}"
+    )
+
+
+def pytest_report_header(config):
+    if random_seeds == RANDOM_SEED_RANGE:
+        seed_values = "all"
+    else:
+        seed_values = ",".join(str(s) for s in random_seeds)
+    return (
+        "To reproduce this test run, set the following environment variable:\n"
+        f'    SKLEARN_TESTS_RANDOM_SEED="{seed_values}"'
+    )
```

Comment:
It doesn't show up in the CI

Reply:
Indeed, an this is really annoying. We cannot merge this PR without fixing this.

It works well locally, even with pytest-xdist enabled so I don't understand why it does not work on the CI.

Reply:
It has to do with running with `--pyargs`. This can be reproduced locally by running:

```bash
pytest --pyargs sklearn.cluster.tests.test_k_means
```

Reply:
Indeed...

Maybe this is related to the second note:

https://docs.pytest.org/en/latest/reference.html#pytest.hookspec.pytest_report_header

Reply:
This is so frustrating. I am not sure what to do. The fixture is running correctly, otherwise the tests would fail. It's just the reporting hook that is ignored.

Reply:
You could add a print in ``test_script.sh`` file that reads the env var. For that you need to set the env var to the generated value when it was not set in the first place. Not the cleanest and a bit sad when there should be a clean way to do that, but meh.

Reply:
Yes this is what I thought... Maybe I'll do that. It's a bit sad.

Reply:
I saw similar issues while trying to get pytest memory profiling (https://github.com/scikit-learn/scikit-learn/pull/21843) to work and pytest + cli option for network tests to work (https://github.com/scikit-learn/scikit-learn/pull/17553).

I suspect we would need to move the code into it's own separate plugin like `sklearn.tests.random_seed` and edit `setup.cfg` to pass `-p sklearn.tests.random_seed` to pytest.

Reply:
I opened https://github.com/ogrisel/scikit-learn/pull/11 to demonstrate the plugin idea.